### PR TITLE
Feature/gap 2060 enhanced block

### PR DESCRIPTION
**Pull request by john-tco**
- adds middleware to validate the admins session token.

Related prs:

User service: https://github.com/cabinetoffice/gap-user-service/pull/113
Front-end: https://github.com/cabinetoffice/gap-find-apply-web/pull/68/files
infra:  https://eu-west-2.console.aws.amazon.com/codesuite/codecommit/repositories/GAP-infra/pull-requests/96/details?region=eu-west-2

### CODE REVIEW

File: src/main/java/gov/cabinetoffice/gap/adminbackend/controllers/UserController.java [hunk 1]
```diff
@@ -29,12 +34,38 @@ public class UserController {
...
+    public ResponseEntity<Boolean> validateAdminSession() {
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+
+        if (authentication == null || !authentication.isAuthenticated()) {
+            return ResponseEntity.ok(Boolean.FALSE);
+        }
+
+        AdminSession adminSession = ((AdminSession) authentication.getPrincipal());
+        if (!oneLoginEnabled) {
+            return ResponseEntity.ok(Boolean.FALSE);
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetoffice/gap/adminbackend/controllers/UserController.java:56]: Why do we invalidate the admin session if one login is not enabled?

File: src/main/java/gov/cabinetoffice/gap/adminbackend/security/JwtTokenFilter.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            return;
+        }
+
+        AdminSession adminSession = ((AdminSession) authentication.getPrincipal());
+        if (!jwtTokenFilterConfig.oneLoginEnabled) {
+            chain.doFilter(request, response);
+            return;
+        }
+
+        String emailAddress = adminSession.getEmailAddress();
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetoffice/gap/adminbackend/security/JwtTokenFilter.java:56]: Need more unit test coverage for this method

File: src/main/java/gov/cabinetoffice/gap/adminbackend/security/JwtTokenFilter.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+
+        if (authentication == null || !authentication.isAuthenticated()) {
+            chain.doFilter(request, response);
+            return;
+        }
+
+        AdminSession adminSession = ((AdminSession) authentication.getPrincipal());
+        if (!jwtTokenFilterConfig.oneLoginEnabled) {
+            chain.doFilter(request, response);
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetoffice/gap/adminbackend/security/JwtTokenFilter.java:52]: We are returning the same as line 46 here?

What happens if One Login is turned off?

File: src/main/java/gov/cabinetoffice/gap/adminbackend/services/UserService.java [hunk 4]
```diff
@@ -29,4 +42,17 @@ public void migrateUser(final String oneLoginSub, final UUID colaSub) {
         });
     }
 
+    public void verifyAdminRoles(final String emailAddress, final String roles) {
+        final String url = userServiceConfig.getDomain() + "/v2/validateAdminSession";
+        final HttpHeaders requestHeaders = new HttpHeaders();
+        requestHeaders.add("emailAddress", emailAddress);
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetoffice/gap/adminbackend/services/UserService.java:48]: I think these should be passed in as query params, any reason why you're passing them in as headers?

File: src/main/java/gov/cabinetoffice/gap/adminbackend/security/JwtTokenFilter.java (hunk 3 above)
Comment (john-tco) [reply src/main/java/gov/cabinetoffice/gap/adminbackend/security/JwtTokenFilter.java]: If onelogin is turned off we will hit chain.doFilter which will skip the verifyRoles call. Users only have roles in after OL migration so we want to let them through with any valid admin session. 

Have added that predicate up & into to the other if statement. tbh the check there for `!authentication.isAuthenticated() `may not even be required as the filter is only applied after the adminsession is verified

File: src/main/java/gov/cabinetoffice/gap/adminbackend/services/UserService.java (hunk 4 above)
Comment (john-tco) [reply src/main/java/gov/cabinetoffice/gap/adminbackend/services/UserService.java]: No good reason, using query params now

File: src/main/java/gov/cabinetoffice/gap/adminbackend/services/UserService.java [hunk 5]
```diff
@@ -29,4 +42,17 @@ public void migrateUser(final String oneLoginSub, final UUID colaSub) {
...
+        requestHeaders.add("roles", roles);
+        final HttpEntity<String> requestEntity = new HttpEntity<>(null, requestHeaders);
+        final Boolean adminSessionIsValid = restTemplate.exchange(url, HttpMethod.GET, requestEntity, Boolean.class)
+                .getBody();
+        if (!adminSessionIsValid) {
```
[the remark this answers is not in the document]
Comment (john-tco) [reply src/main/java/gov/cabinetoffice/gap/adminbackend/services/UserService.java]: it seems to work with both - have swtiched to that

[the remark this answers is not in the document]
Comment (dylanwrightCO) [reply src/main/java/gov/cabinetoffice/gap/adminbackend/services/UserService.java]: What happens if adminSessionIsValid is null?

Should add a check in the conditional to throw the exception:

```
 if (adminSessionIsValid == null || adminSessionIsValid.equals(Boolean.FALSE)) {
            throw new UnauthorizedException("Token is not valid");
        }
```

File: src/main/java/gov/cabinetoffice/gap/adminbackend/security/JwtTokenFilter.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+import static org.springframework.util.ObjectUtils.isEmpty;
+
+/**
+ * This class cannot be a Spring bean, otherwise Spring will automatically apply it to all
+ * requests, regardless of whether they've been specifically ignored
+ */
+@RequiredArgsConstructor
+public class JwtTokenFilter extends OncePerRequestFilter {
+
+    private final JwtService jwtService;
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetoffice/gap/adminbackend/security/JwtTokenFilter.java:35]: Is this used or required?

File: src/main/java/gov/cabinetoffice/gap/adminbackend/controllers/UserController.java [hunk 7]
```diff
@@ -29,12 +34,38 @@ public class UserController {
...
+        }
+        String emailAddress = adminSession.getEmailAddress();
+        String roles = adminSession.getRoles();
+        try {
+            userService.verifyAdminRoles(emailAddress, roles);
+            return ResponseEntity.ok(Boolean.TRUE);
+        }
+        catch (UnauthorizedException error) {
+            return ResponseEntity.ok(Boolean.FALSE);
+        }
```
Comment (westdom) [inline src/main/java/gov/cabinetoffice/gap/adminbackend/controllers/UserController.java:66]:
```suggestion
        return userService.verifyAdminRoles(emailAddress, roles);
```

File: src/main/java/gov/cabinetoffice/gap/adminbackend/services/UserService.java [hunk 8]
```diff
@@ -29,4 +42,15 @@ public void migrateUser(final String oneLoginSub, final UUID colaSub) {
...
+    public void verifyAdminRoles(final String emailAddress, final String roles) {
+        final String url = userServiceConfig.getDomain() + "/v2/validateAdminSession?emailAddress=" + emailAddress
+                + "&roles=" + roles;
+        final HttpEntity<String> requestEntity = new HttpEntity<>(null);
+        final Boolean adminSessionIsValid = restTemplate.exchange(url, HttpMethod.GET, requestEntity, Boolean.class)
+                .getBody();
+        if (adminSessionIsValid.equals(Boolean.FALSE)) {
+            throw new UnauthorizedException("Token is not valid");
+        }
+    }
```
Comment (westdom) [inline src/main/java/gov/cabinetoffice/gap/adminbackend/services/UserService.java:54]:
```suggestion
    public boolean verifyAdminRoles(final String emailAddress, final String roles) {
        final String url = userServiceConfig.getDomain() + "/v2/validateAdminSession?emailAddress=" + emailAddress
                + "&roles=" + roles;
        final HttpEntity<String> requestEntity = new HttpEntity<>(null);
        return restTemplate.exchange(url, HttpMethod.GET, requestEntity, Boolean.class)
                .getBody();
    }
```

File: src/main/java/gov/cabinetoffice/gap/adminbackend/services/UserService.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (john-tco) [reply src/main/java/gov/cabinetoffice/gap/adminbackend/services/UserService.java]: added a null check

File: src/main/java/gov/cabinetoffice/gap/adminbackend/services/UserService.java (hunk 8 above)
Comment (john-tco) [reply src/main/java/gov/cabinetoffice/gap/adminbackend/services/UserService.java]: Nice, thanks - have added that

File: src/main/java/gov/cabinetoffice/gap/adminbackend/security/JwtTokenFilter.java (hunk 6 above)
Comment (john-tco) [reply src/main/java/gov/cabinetoffice/gap/adminbackend/security/JwtTokenFilter.java]: removed

File: src/main/java/gov/cabinetoffice/gap/adminbackend/services/UserService.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (dylanwrightCO) [reply src/main/java/gov/cabinetoffice/gap/adminbackend/services/UserService.java]: Noticed you are no longer throwing an exception if isAdminSessionValid is FALSE.

If the admin session cannot be validated and therefore FALSE is returned to the client, does it get handled properly?

[the remark this answers is not in the document]
Comment (john-tco) [reply src/main/java/gov/cabinetoffice/gap/adminbackend/services/UserService.java]: false doesn't get returned to the client anymore, we jsut throw unauthorised in all cases (in validateSessionRoles). On the front-end we check that the body is exactly true.

const isAdminSessionValid = async (sessionCookie: string) => {
  const response = await fetch(`${BASE_USERS_URL}/validateAdminSession`, {
    headers: {
      Cookie: `SESSION=${sessionCookie};`,
    },
  });
  return (await response.text()) === 'true';
};

**Review by dominicwest — APPROVED** [review]

File: src/main/java/gov/cabinetoffice/gap/adminbackend/security/WebSecurityConfig.java [hunk 9]
```diff
@@ -9,37 +12,46 @@
...
+                                "/users/migrate")
+                        .permitAll()
+                        .antMatchers("/v3/api-docs/**", "/swagger-ui/**", "/swagger-resources/**", "/swagger-ui.html",
+                                "/webjars/**")
+                        .permitAll().anyRequest().authenticated())
 
                 .formLogin().disable().httpBasic().disable().logout().disable().csrf().disable().exceptionHandling()
                 .authenticationEntryPoint(new HttpStatusEntryPoint(HttpStatus.UNAUTHORIZED));
+
+        http.addFilterAfter(jwtTokenFilter, UsernamePasswordAuthenticationFilter.class);
```
Comment (dominicwest) [inline src/main/java/gov/cabinetoffice/gap/adminbackend/security/WebSecurityConfig.java:53]: Could feature flag this line here but passing it into the filter also works

**Pull request merged**